ride/lp_stable.ride

```diff
@@ -533,6 +550,203 @@ func cp(caller: String,
+
+  strict checkD = D1 > D0 || throw()
+
+  let lpAssetEmission = lpId.fromBase58String().assetInfo().valueOrErrorMessage("invalid lp asset").quantity.toBigInt()
+  let lpAmount = if (lpAssetEmission > big0) then {
+    fraction(lpAssetEmission, D1 - D0, D0)
+  } else "initial deposit requires all coins".throw()
+
+  let curPrice = cpbi(prBalance.t1(priceDcm), amBalance.t1(amtDcm)).f1(scale8)
+
+  let commonState = [
+    IntegerEntry(pl(), curPrice),
+    IntegerEntry(ph(height, lastBlock.timestamp), curPrice),
+    StringEntry(
+      pau(userAddress, txId),
+      dataPutActionInfo(amAmount, prAmount, lpAmount.toInt(), curPrice, 0, 0, height, lastBlock.timestamp, 0, 0)
+    )
+  ]
+
+  let poolProportion = prBalanceOld.fraction(scale8, amBalanceOld)
+  let amountAssetPart = fraction(pmtAmt, scale8, poolProportion + scale8)
+  let priceAssetPart = pmtAmt - amountAssetPart
+  let lpAmtBoth = fraction(lpAssetEmission, priceAssetPart.toBigInt(), prBalanceOld.toBigInt())
+  let bonus = fraction(lpAmount - lpAmtBoth, scale8BigInt, lpAmtBoth).toInt()
+  
+  (lpAmount.toInt(), commonState, feeAmount, bonus)
+}
+
+func getOneTknV2Internal(
+  outAssetId: String,
+  minOutAmount: Int,
+  payments: List[AttachedPayment],
+  caller: Address,
+  originCaller: Address,
+  transactionId: ByteVector
+) = {
+  if (payments.size() != 1) then throw("1 pmnt expd") else
+  let poolConfig = gpc()
+  let lpId = poolConfig[idxLPAsId]
+  let amId = poolConfig[idxAmAsId]
+  let prId = poolConfig[idxPrAsId]
+  let amDecimals = poolConfig[idxAmtAsDcm].parseIntValue()
+  let prDecimals = poolConfig[idxPriceAsDcm].parseIntValue()
+  let poolStatus = poolConfig[idxPoolSt]
+
+  let userAddress = if (caller == this) then originCaller else caller
+
+  let pmt = payments[0].value()
+  let pmtAssetId = pmt.assetId.value()
+  let pmtAmt = pmt.amount
+
+  if (pmtAmt < 10_00_000_000) then throw("Min pmt 10 LP") else
+
+  let txId58 = transactionId.toBase58String()
+
+  if (lpId != pmtAssetId.toBase58String()) then throw("Wrong LP") else
+
+  let amBalance = amId.getAccBalance()
+  let prBalance = prId.getAccBalance()
+
+  let (totalGet, feeAmount) = this.invoke("getOneTknV2READONLY", [outAssetId, pmtAmt], []).exactAs[(Int, Int)]
+
+  let totalAmount = if (minOutAmount > 0 && totalGet < minOutAmount) then {
+    ["amount to receive is less than ", minOutAmount.toString()].makeString("").throw()
+  } else totalGet
+
+  let (outAm, outPr) = if(outAssetId == amId) then (totalAmount, 0) else (0, totalAmount)
+
+  let curPrX18 = cpbi(prBalance.t1(prDecimals), amBalance.t1(amDecimals))
+  let curPr = curPrX18.f1(scale8)
+
+  let outAssetIdOrWaves = if (outAssetId == "WAVES") then unit else outAssetId.fromBase58String()
+  let sendFeeToMatcher = if (feeAmount > 0) then [ScriptTransfer(feeCollectorAddress, feeAmount, outAssetIdOrWaves)] else []
+
+  strict state = [
+    ScriptTransfer(userAddress, totalAmount, outAssetIdOrWaves),
```

Comment:
HIGH: Not enough `outAssetId` validation can lead to `totalAmount` transfer of any token != `amId`, should be fully validated in this places:
https://github.com/waves-exchange/contracts/blob/171daaee20cab6ad3c618ef50662e04172cd4d45/ride/lp_stable.ride#L1150
https://github.com/waves-exchange/contracts/blob/171daaee20cab6ad3c618ef50662e04172cd4d45/ride/lp_stable.ride#L725